File: lib/Mod_Comb.py
MOD = 10**9+7
MAX = 3*(10**5)+2

fact, inverse = [0]*(MAX), [0]*(MAX)
fact[0], inverse[0] = 1, 1
for i in range(1, MAX):
  fact[i] = (fact[i-1]*i)%MOD
  inverse[i] = pow(fact[i], MOD-2, MOD)

def mod_comb(n, r):
  if n<r or n==0: return 0
  if r==0: return 1
  return (((fact[n]*inverse[r])%MOD)*inverse[n-r])%MOD

File: Lib/C++/mod_comb.cpp
const int MAX = 210000;
const int MOD = 1000000007;

long long fact[MAX], inv[MAX], invfact[MAX];
void comb_build() {
  fact[0] = fact[1] = 1;
  inv[1] = 1;
  invfact[0] = 0, invfact[1] = 1;
  for (int i = 2; i < MAX; i++) {
    fact[i] = fact[i-1]*i%MOD;
    inv[i] = MOD - inv[MOD%i]*(MOD/i)%MOD;
    invfact[i] = invfact[i-1]+inv[i]%MOD;
  }
}

long long nCk(int n, int k){
  if(n<0||k<0||k>n) return 0;
  return (fact[n]*invfact[k]%MOD)*invfact[n-k]%MOD;
}